src/Castle.Core/DynamicProxy/Generators/GeneratorUtil.cs

```diff
@@ -33,13 +35,55 @@ public static class GeneratorUtil
 
 			var arguments = StoreInvocationArgumentsInLocal(emitter, invocation);
 
+			// Extracted from the following `for` loop as an optimization.
+			//
+			// The comparison by name is intentional; any assembly could define that attribute.
+			// See explanation in comment below.
+			Func<object, bool> isIsReadOnlyAttribute =
+				attribute => attribute.GetType().FullName == "System.Runtime.CompilerServices.IsReadOnlyAttribute";
+
 			for (var i = 0; i < parameters.Length; i++)
 			{
 				if (!parameters[i].ParameterType.GetTypeInfo().IsByRef)
 				{
 					continue;
 				}
 
+				// C# `in` parameters are also by-ref, but meant to be read-only.
+				// The section "Metadata representation of in parameters" on the following page
+				// defines how such parameters are marked:
+				//
+				// https://github.com/dotnet/csharplang/blob/master/proposals/csharp-7.2/readonly-ref.md
+				//
+				// This poses three problems for detecting them:
+				//
+				//  * The C# Roslyn compiler marks `in` parameters with an `[in]` IL modifier,
+				//    but this isn't specified, nor is it used uniquely for `in` params.
+				//
+				//  * `System.Runtime.CompilerServices.IsReadOnlyAttribute` is not defined on all
+				//    .NET platforms, so the compiler sometimes recreates that type in the same
+				//    assembly that contains the method having an `in` parameter. In other words,
+				//    it's an attribute one must check for by name (which is slow, as it implies
+				//    use of a `GetCustomAttributes` enumeration instead of a faster `IsDefined`).
+				//
+				//  * A required custom modifier `System.Runtime.InteropServices.InAttribute`
+				//    is always present in those cases relevant for DynamicProxy (proxyable methods),
+				//    but not all targeted platforms support reading custom modifiers. Also,
+				//    support for cmods is generally flaky (at this time of writing, mid-2018).
+				//
+				// The above points inform the following detection logic: First, we rely on an IL
+				// `[in]` modifier being present. This is a "fast guard" for non-`in` parameters:
+				if ((parameters[i].Attributes & ParameterAttributes.In) != 0)
```

Comment:
As to me, this method started to have too many responsibilities. Better create a standalone method like `ArgumentsUtil.IsReadOnlyByRefArg()` or similar and put all the detection logic there.

Also it might make sense to create `ArgumentsUtil.IsAnyWritableByRef` or similar and use it instead of `ArgumentsUtil.IsAnyByRef`, so we don't generate e.g. redundant `try/catch` when we are not going to write something.

Reply:
Given that the method is called `CopyOutAndRefParameters`, I think the additional 6 lines of code are still in scope.

That said, I do agree that it could be re-written to make better use of small helper methods to somewhat raise the level of abstraction.

You'll note there's already a method `AttributeUtil.IsAnyByRef`, so `AttributeUtil` might be a good home for additional similar methods such as `bool IsIn(ParameterInfo)`, `bool IsRef(ParameterInfo)`, `bool IsOut(ParameterInfo)` or just one `ParameterByRefKind GetParameterByRefKind(ParameterInfo)`. But once we start adding methods there, we also have to talk about accessibility for consistency's sake. The present-day `AttributeUtil.IsAnyByRef` is `public` which I see as a mistake. So either we expand Core's public API further, or we add methods with inconsistent accessibility, or we add the method in a different place, which is also somewhat chaotic.

This is a design problem I'd rather not solve today. May I leave it up to you to submit a PR that cleans this method up a little?

Reply:
Well, if you don't want to produce too much noise, then I'd suggest to simply extract all the `IsReadOnlyRef` logic to a static method in the same class. Each method should have single responsibility only, while currently `CopyOutAndRefParameters` method orchestrates parameters handling and knows the particularity of how `in` modifier is implemented.